README.md
Dymo
====

Dymo is a placement script for map labels, isolated from the purpose-built code in
[GeoIQ’s Acetate](https://github.com/fortiusone/acetate). Dymo resolves positions for
densely-packed point labels, and results in layouts make your maps look like they’ve been
[touched by a cartographer](http://www.flickr.com/photos/mmigurski/5194528400/sizes/l/in/photostream/).

Dymo uses [Simulated Annealing](http://en.wikipedia.org/wiki/Simulated_annealing)
to derive an acceptable global label set, described in Steven Wouderberg’s
[October 2007 presentation](http://www.cs.uu.nl/docs/vakken/gd/steven2.pdf)
at Utrecht University. See an animation of the process over time in these two
videos of U.S. and European place names:

<p>
    <a href="http://vimeo.com/migurski/dymo-n-america"><img src="https://github.com/migurski/Dymo/raw/master/images/us-video-still.jpg" width="440" height="219"></a>
    <a href="http://vimeo.com/migurski/dymo-europe"><img src="https://github.com/migurski/Dymo/raw/master/images/europe-video-still.jpg" width="440" height="219"></a>
</p>

QUICK START
----------

**Demo area:** Run `make geojson` to anneal labels for San Francisco and Israel sample areas from zooms 4 to 10. This will automatically build the data files, too (eg: `make data`). Should run in less than an hour.

**World:** Annealing the entire world will take several days to several weeks and will require lots of RAM and many processor cores. Modify the `Makefile` to remove the spatial filter as below:

    # COMPOSITE FILTER (fast!)
    # SPATIAL_FILTER=  --filter-bounding-box $(FILTER_SANFRANCISCO) --filter-bounding-box $(FILTER_JERUSALEM)
    # NO FILTER (slow! If no filter is desired, uncomment line below, comment out line above.)
    SPATIAL_FILTER=

You can change the fonts and population steps in the Makefile, too.

DETAILS
----------

```dymo-label.py``` is a script that converts lists of cities with included font
information to GeoJSON point and label files suitable for use in geographic
rendering.

Mamata Akella at the National Park Service has written [a detailed tutorial on Dymo and Tile Mill](http://www.nps.gov/npmap/blog/improving-park-label-placement-with-dymo-and-tilemill.html), to “avoid label overlaps and improve the overall legibility of park names dramatically.”
The tutorial shows how to prepare data for Dymo using ArcGIS Desktop.

###Label Usage

Place U.S. city labels at zoom 6 for two minutes:

    python dymo-label.py -z 6 --minutes 2 --labels-file labels.json --places-file points.json data/US-z6.csv.gz

Place U.S. city labels at zoom 5 over a 10000-iteration 10.0 - 0.01 temperature range:

    python dymo-label.py -z 5 --steps 10000 --max-temp 10 --min-temp 0.01 -l labels.json -p points.json data/US-z5.csv

Both examples will result in a pair of GeoJSON files, ```labels.json``` and
```points.json.``` The first will contain rectangular label areas, the second
will contain center points of places successfully positioned by Dymo. Because
labels will collide in different ways depending on map scale, labels must be
placed separately for each zoom level:

![U.S. Labels](http://farm5.static.flickr.com/4123/5413923643_be14a6b13b_b.jpg)

For larger datasets, it’s almost always faster to run `dymo-label.py` with the
`--minutes` option instead of `--min-temp`/`--max-temp`, because Dymo will
automatically partition labels based on mutual overlaps and perform many small
annealing processes.

Data Included
-------------

Look in ```data/``` for a list of zoom-by-zoom city locations, organized by
continent and selected by Justin O’Bierne. Data is distributed separately from
code, see [downloads](https://github.com/migurski/Dymo/downloads) for a link.

To prepare your own city lists or modify fonts and font sizes in input lists,
Use ```dymo-prepare-places.py``` to apply population-specific font choices to
an in-bound list.

###Prepare-Places Usage

    python dymo-prepare-places.py --font 0 fonts/Arial.ttf 10 --font 1000000 fonts/Arial.ttf 12 --zoom 5 data/North-America-all.txt.gz data/North-America-z5.txt

Sample Output
-------------

These sample images were created during the development of Acetate, and show
the results of town placement.

[Central Japan](http://www.flickr.com/photos/mmigurski/5194528400/):

![Central Japan](http://farm5.static.flickr.com/4148/5194528400_abf78d0c09_b.jpg)

[Western Europe](http://www.flickr.com/photos/mmigurski/5193928891/):

![Western Europe](http://farm5.static.flickr.com/4111/5193928891_25ae4e213c_b.jpg)

[Northen Appalachians](http://www.flickr.com/photos/mmigurski/5194532290/):

![Northen Appalachians](http://farm5.static.flickr.com/4092/5194532290_96776a8241_b.jpg)

Advanced Options
----------------

###Dymo Prepare Places

* `-z ZOOM, --zoom=ZOOM` - Maximum zoom level. _Default value is `4`._
* `--zoom-field=ZOOM_FIELD` - Field to use for limiting selection by zoom. _Default field is `zoom start`._
* `-f FONTS, --font=FONTS` - Additional font, in the form of three values: minimum population number (or other font field), font file path, font size. Can be specified multiple times.
* `-r RADIUS, --radius=RADIUS` - Pixel buffer around each place. _Default value is `0`._
* `--font-field=FONT_FIELD` - Field to use for font selection. _Default field is `population`._
* `--filter-field=FILTER_FIELD` - Field to use for limiting selection by theme and the value to limit by. _Default is no filter._
* `--symbol-size=SYMBOL_SIZE` - Size in pixels for implied townspot symbol width/height in pixels. _Default size is `8`._
* `--symbol-size-field=SYMBOL_SIZE_FIELD` - Field to use for sizing the implied townspot symbol width/height in pixels. _No default._

###Dymo Label

* `-m MINUTES, --minutes=MINUTES` - Number of minutes to run annealer. Longer run times generally give better results, to a point. _Default value is `2.0`._
* `-z ZOOM, --zoom=ZOOM` -  Map zoom level. Conflicts with `--scale` and `--projection` options. _Default value is `18`._
* `-l LABELS_FILE, --labels-file=LABELS_FILE` - _Optional: name of labels file to generate._
* `-p PLACES_FILE, --places-file=PLACES_FILE` - _Optional: name of place points file to generate._
* `-r REGISTRATIONS_FILE, --registrations-file=REGISTRATIONS_FILE` - _Optional: name of registration points file to generate._ This file will have an additional "justified" field with values "left", "center", or "right".
* `--min-temp=TEMP_MIN` - Optional: Minimum annealing temperature, for more precise control than specifying --minutes.
* `--max-temp=TEMP_MAX` - Optional: Maximum annealing temperature, for more precise control than specifying --minutes.
* `--steps=STEPS` - Number of annealing steps, for more precise control than specifying --minutes.
* `--include-overlaps` - Include lower-priority places when they overlap higher-priority places. _Default behavior is to skip the overlapping cities._
* `--output-projected` - Optional: output projected coordinates.
* `--projection=PROJECTION` - Optional: PROJ.4 string to use instead of default web spherical mercator.
* `--blobs` - Load input as blobs rather than points, placing labels on top of locations instead of near them.
* `--scale=SCALE` - Scale to use with --projection. Equivalent to +to_meter PROJ.4 parameter, which is not used internally due to not quite working in pyproj. Conflicts with --zoom option. _Default value is `1`._
* `--dump-file=DUMP_FILE` - Optional: filename for a sequential dump of pickled annealer states. This all has to be stored in memory, so for a large job specifying this option could use up all available RAM.
* `--dump-skip=DUMP_SKIP` - Optional: number of states to skip for each state in the dump file.
* `--name-field=NAME_FIELD` - Name of column for labels to name themselves. _Default value is `name`._
* `--placement-field=PLACEMENT_FIELD` - Optional: name of column for point placement. Default value is `preferred placement`.

###Advanced Prepare Places Usage

* Use larger or smaller townspot symbol sizes: `--symbol-size`

        python dymo-prepare-places.py --symbol-size 6 --font 0 fonts/Arial.ttf 10 --font 1000000 fonts/Arial.ttf 12 --zoom 5 data/North-America-all.txt.gz data/North-America-z5.txt

* Use custom townspot symbol size per feature: `--symbol-size-field`

        python dymo-prepare-places.py --symbol-size-field spotsize --font 0 fonts/Arial.ttf 10 --font 1000000 fonts/Arial.ttf 12 --zoom 5 data/North-America-all.txt.gz data/North-America-z5.txt

* Use a custom 'population' attribute to size grade the text labels: `--font-field`

        python dymo-prepare-places.py --font-field rank --font 0 fonts/Arial.ttf 10 --font 1000000 fonts/Arial.ttf 12 --zoom 5 data/North-America-all.txt.gz data/North-America-z5.txt

* Limit the initial visibility of a feature to a specific zoom: `--zoom-field`

        python dymo-prepare-places.py --zoom-field init_zoom --font 0 fonts/Arial.ttf 10 --font 1000000 fonts/Arial.ttf 12 --zoom 5 data/North-America-all.txt.gz data/North-America-z5.txt

* Subselect features based on a simple filter: `--filter-field`

        python dymo-prepare-places.py --filter-field city_type capital_city --font 0 fonts/Arial.ttf 10 --font 1000000 fonts/Arial.ttf 12 --zoom 5 data/North-America-all.txt.gz data/North-America-z5.txt

* Combo:

        python dymo-prepare-places.py -z 6 --radius 1 --font 0 "fonts/Arial.ttf" 14 --font 4 "fonts/Arial Bold.ttf" 18 --font-field "sizeclass" master_data_file.tsv --zoom-field "zoom_start" --symbol-size-field "symbol_size" --filter-field region west west-labels-z6.txt


###Advanced Label Usage

* Run longer, look prettier: adjust `--minutes`, `--min-temp`, `--max-temp`, and `--steps`
* Use a custom `name` field: specify a custom `--name-field`

        python dymo-label.py --name-field name_ascii -z 6 --minutes 2 --labels-file labels.json --places-file points.json data/North-America-z5.txt

* Output multiple label, places, registration outputs: Use `--labels-file`, `--places-file`, and `--registrations-file`.

        python dymo-label.py --registrations-file registrations.json -z 6 --minutes 2 --labels-file labels.json --places-file points.json data/North-America-z5.txt

* Provide label hints to the auto-label algorithm: Use `--placement-field`.

        python dymo-label.py --placement-field hint -z 6 --minutes 2 --labels-file labels.json --places-file points.json data/North-America-z5.txt

* Manually troubleshoot overlaps in QGIS: Use `--include-overlaps` and open in QGIS and fine tune output.

        python dymo-label.py --include-overlaps -z 6 --minutes 2 --labels-file labels.json --places-file points.json data/North-America-z5.txt

* Use a custom map projection: Use `--output-projected`, `--projection`, and `--scale`

    For an Albers map of the USA, but saved in WGS84 geographic coodinates:

        python dymo-label.py --projection "+proj=aea +lat_1=29.5 +lat_2=45.5 +lat_0=38.0 +lon_0=-98.0 +ellps=WGS84 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs" --scale 2377 --steps 50000 --max-temp 230 --min-temp 0.0001 --labels-file northeast-labels.json east-labels-z3.txt --name-field "label" --placement-field "preferred_z3"
        
    For an Albers map of the USA, saved in that projection:

        python dymo-label.py --projection "+proj=aea +lat_1=29.5 +lat_2=45.5 +lat_0=38.0 +lon_0=-98.0 +ellps=WGS84 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs" --output-projected --scale 2377 --steps 50000 --max-temp 230 --min-temp 0.0001 --labels-file northeast-labels.json east-labels-z3.txt --name-field "label" --placement-field "preferred_z3"


Tips
----

* **Converting to Esri Shapefile format:** Use [GDAL/OGR's]() `ogr2ogr` tool to convert the GeoJSON Dymo output. [QGIS](), a free desktop GIS application, will also convert the files to SHP. 

		ogr2ogr -f 'Esri Shapefile' -lco=UTF8 output.shp input.json
		
    If you have OGR 1.9+ (for the UTF8 layer creation option to preserve the unicode place names), run `make shp` from the Dymo directory and the SHP versions will be automatically created.

* **Reduce the file size** of the GeoJSON output using [Lil'JSON](https://github.com/migurski/LilJSON/):

        python LilJSON.py --precision 3 input.json output.json


Who
---

Copyright 2010-2012 Michal Migurski, Nathaniel V. Kelso, and GeoIQ, offered under the [BSD license](http://www.opensource.org/licenses/bsd-license.php). Uses Richard J. Wagner’s [Python annealing library](http://www-personal.umich.edu/~wagnerr/PythonAnneal.html).

We’re not affiliated with [Dymo Corporation](http://dymo.com).


CHANGELOG
2012-10-05: 0.16.0
- Added new world data file with new continent attributes on places
- Removed expired per continent data files
- Updated Makefile to use new data files
- Updated Makefile to use small sample data (San Francisco and Israel) via newer spatial filter on prepare places.
- Updated Makefile to also convert GeoJSON output to SHP files (optional make target)
- Added GeoJSON demo output (San Francisco and Israel)
- Added SHP shapefile demo output (San Francisco and Israel)
- Added Quick Start section to Readme with instructions on `make geojson` and `make shp` targets.

2012-10-05: 0.15.0
- Added --append flag to dymo-label for appending to output files.

2012-09-24: 0.14.2
- Modified internal multiprocessing use so that ctrl-C correctly stops dymo-label.

2012-09-24: 0.14.1
- Fixed handling of dataets that require no annealing.

2012-09-24: 0.14.0
- Added spatial filtering for multiple bounding boxes in dymp-prepare-places.
- Added wider variety of allowed lat/lon field names for spatial queries.
- Fixed empty file output.

2012-09-23: 0.13.0
- Added multiprocessing option for dymo-label.
- Added more verbose/quiet command line flags.
- Fixed blobs mode.

2012-09-14: 0.12.0 (after the fact 2012-01-17)
- Added support for arbitrary place label text in --name-field option.
- Added support for arbitrary "zoom start" --zoom-field option.
- Added support for variable townspot symbol point sizes with --symbol-size (universal) and --symbol-size-field (per feature) options.
- Added support to limit results based on --filter-field value match.

2012-06-18: 0.11.2
- Modified partition weighting to actual instead of potential connectivity.

2012-06-10: 0.11.1
- Fixed a namespace conflict between Dymo Point and shapely Point.

2012-06-10: 0.11.0
- Added --blobs option to dymo-label.py for making non-town-point output.

2012-06-09: 0.10.1
- Fixed a dumb mistake in setup.py.

2012-06-09: 0.10.0
- Added partitioning behavior to dymo-label for speed boost & potential concurrency.

2012-05-20: 0.9.1
- Fixed shbang so that scripts actually work when installed.

2012-01-10: 0.9.0
- Added support for arbitrary projections in dymo-label.py with new --projection and --scale options.
- Allowing latitude and longitude to be specified as "lat", "lon", or "long", upper- or lowercase.

2011-11-18: 0.8.1
- Added simple bounds check for latitude and longitude of places.

2011-10-01: 0.8.0
- Incompatibly-modified dymo-label.py syntax with --labels-file, --places-file and --registrations-file options.
- Added text registration point and justification output.
- Added preferred placement support for point labels.

2011-09-29: 0.7.2
- Using quadtree indexes for fast collision checks in labeling and preparation.

2011-09-28: 0.7.1
- Data is now distributed in a separate tarball.

2011-09-28: 0.7
- Added --radius option to dymo-prepare-place.py to cull tightly-packed places.
- Added --font-field option to dymo-prepare-place.py to use fields other than population.

2011-09-23: 0.6
- Added awareness of point size column to dymo-prepare-places.py.

2011-09-23: 0.5
- Using a higher-precision guess for font size in places.Place.

2011-09-23: 0.4
- Everything still works when there's nothing to do.

2011-09-23: 0.3
- Changed some assumptions about column names, e.g. "zoom start" instead of "zoom".

2011-09-23: 0.2
- Fixed unicode / utf-8 decode issue in load_places()


dymo-label.py
#!/usr/bin/env python

from optparse import OptionParser
from multiprocessing import Process, Queue, JoinableQueue
from copy import copy, deepcopy
from datetime import timedelta
from os.path import exists
from Queue import Empty
from time import time
import logging
import cPickle
import json

from Dymo.anneal import Annealer
from Dymo.index import FootprintIndex
from Dymo.places import Places, NothingToDo
from Dymo import load_places, load_blobs, get_geometry

def anneal_placelist(places, indexes, weight, connections):
    ''' Anneal a list of places and return the results.
    '''
    if len(indexes) == 1:
        return [(indexes[i], place) for (i, place) in enumerate(places)]
    
    logging.info('Placing '+', '.join(sorted([place.name.encode('utf-8', 'replace') for place in places])))

    try:
        start = time()
        minutes = options.minutes * float(weight) / connections
        places, e = annealer.auto(places, minutes, min(100, weight * 20))

    except NothingToDo:
        pass
    
    else:
        if minutes > .3:
            elapsed = timedelta(seconds=time() - start)
            overtime = elapsed - timedelta(minutes=minutes)
            logging.debug('...done in %s including %s overhead.' % (str(elapsed)[:-7], str(overtime)[:-7]))
    
    return [(indexes[i], place) for (i, place) in enumerate(places)]

def anneal_in_serial(places):
    '''
    '''
    annealed = [None] * places.count()
    
    for piece in places.in_pieces():
        for (index, place) in anneal_placelist(*piece):
            assert annealed[index] is None
            annealed[index] = place
    
    return annealed

def anneal_in_parallel(places, processes):
    '''
    '''
    def do_queue(src_queue, dst_queue):
        ''' Pull tuples from src_queue, put results of anneal_placelist() onto dst_queue.
        '''
        while True:
            places, indexes, weight, connections = src_queue.get(True)
            annealed_places = anneal_placelist(places, indexes, weight, connections)
            
            for (index, place) in annealed_places:
                dst_queue.put((index, place))
    
            src_queue.task_done()
    
    annealed = [None] * places.count()
    
    src_queue, dst_queue = JoinableQueue(), Queue()
    processes = [Process(target=do_queue, args=(src_queue, dst_queue)) for i in range(processes)]
    
    # load up a source queue
    for (places, indexes, weight, connections) in places.in_pieces():
        src_queue.put((places, indexes, weight, connections))
    
    # start all the workers
    for proc in processes:
        proc.start()
    
    src_queue.join()
    
    # grab all the results
    while True:
        try:
            index, place = dst_queue.get(True, .1)
            assert annealed[index] is None
            annealed[index] = place
        except Empty:
            break
    
    # stop all the workers
    for proc in processes:
        proc.terminate()
    
    return annealed

optparser = OptionParser(usage="""%prog [options] --labels-file <label output file> --places-file <point output file> --registrations-file <registration output file> <input file 1> [<input file 2>, ...]

There are two ways to run the label placer. The default way performs a test
to figure out the best parameters for the simulated annealing algorithm before
running it. The more precise way required that you know what your minimum and
maximum temperatures and appropriate number of steps are before you start.
For most inputs, the default method with `--minutes` specified will be best.

Input fields:

  preferred placement
    Optional preference for point placement, one of "top right" (the default),
    "top", "top left", "bottom left", "bottom", or "bottom right". The name of this
    field can be changed using the --placement-field option.

Examples:

  Place U.S. city labels at zoom 6 for two minutes:
  > python dymo-label.py -z 6 --minutes 2 --labels-file labels.json --places-file points.json data/US-z6.csv.gz

  Place U.S. city labels at zoom 5 over a 10000-iteration 10.0 - 0.01 temperature range:
  > python dymo-label.py -z 5 --steps 10000 --max-temp 10 --min-temp 0.01 -l labels.json -p points.json data/US-z5.csv""")

defaults = dict(minutes=2, dump_skip=100, include_overlaps=False, output_projected=False, load_inputs=load_places, name_field='name', placement_field='preferred placement', append=False, processes=1, verbose=None)

optparser.set_defaults(**defaults)

optparser.add_option('-m', '--minutes', dest='minutes',
                     type='float', help='Number of minutes to run annealer. Default value is %(minutes).1f.' % defaults)

optparser.add_option('-z', '--zoom', dest='zoom',
                     type='int', help='Map zoom level. Conflicts with --scale and --projection options. Default value is 18.' % defaults)

optparser.add_option('-l', '--labels-file', dest='labels_file',
                     help='Optional name of labels file to generate.')

optparser.add_option('-p', '--places-file', dest='places_file',
                     help='Optional name of place points file to generate.')

optparser.add_option('-r', '--registrations-file', dest='registrations_file',
                     help='Optional name of registration points file to generate. This file will have an additional "justified" field with values "left", "center", or "right".')

optparser.add_option('--min-temp', dest='temp_min',
                     type='float', help='Minimum annealing temperature, for more precise control than specifying --minutes.')

optparser.add_option('--max-temp', dest='temp_max',
                     type='float', help='Maximum annealing temperature, for more precise control than specifying --minutes.')

optparser.add_option('--steps', dest='steps',
                     type='int', help='Number of annealing steps, for more precise control than specifying --minutes.')

optparser.add_option('--include-overlaps', dest='include_overlaps',
                     action='store_true', help='Include lower-priority places when they overlap higher-priority places. Default behavior is to skip the overlapping cities.')

optparser.add_option('--output-projected', dest='output_projected',
                     action='store_true', help='Optionally output projected coordinates.')

optparser.add_option('--projection', dest='projection',
                     help='Optional PROJ.4 string to use instead of default web spherical mercator.')

optparser.add_option('--blobs', dest='load_inputs', action='store_const', const=load_blobs,
                     help='Load input as blobs rather than points, placing labels on top of locations instead of near them.')

optparser.add_option('--scale', dest='scale',
                     type='float', help='Optional scale to use with --projection. Equivalent to +to_meter PROJ.4 parameter, which is not used internally due to not quite working in pyproj. Conflicts with --zoom option. Default value is 1.')

optparser.add_option('--dump-file', dest='dump_file',
                     help='Optional filename for a sequential dump of pickled annealer states. This all has to be stored in memory, so for a large job specifying this option could use up all available RAM.')

optparser.add_option('--dump-skip', dest='dump_skip',
                     type='int', help='Optional number of states to skip for each state in the dump file.')

optparser.add_option('--name-field', dest='name_field',
                     help='Optional name of column for labels to name themselves. Default value is "%(name_field)s".' % defaults)

optparser.add_option('--placement-field', dest='placement_field',
                     help='Optional name of column for point placement. Default value is "%(placement_field)s".' % defaults)

optparser.add_option('--append', dest='append',
                     action='store_true', help='Append to existing output files instead of overwriting them.')

optparser.add_option('-P', '--processes', dest='processes',
                     type='int', help='Number of concurrent annealing processes to run. Default value is %(processes)d.' % defaults)

optparser.add_option('-v', '--verbose', dest='verbose',
                     action='store_true', help='Be extra chatty when running.')

optparser.add_option('-q', '--quiet', dest='verbose',
                     action='store_false', help='Be extra quiet when running.')


if __name__ == '__main__':
    
    options, input_files = optparser.parse_args()
    
    if options.verbose:
        log_level = logging.DEBUG
    elif options.verbose is False:
        log_level = logging.WARNING
    else:
        log_level = logging.INFO
    
    logging.basicConfig(format='%(msg)s', level=log_level)
    
    #
    # Geographic projections
    #
    
    if options.zoom is not None and options.scale is not None:
        logging.critical('Conflicting input: --scale and --zoom can not be used together.')
        exit(1)
    
    if options.zoom is not None and options.projection is not None:
        logging.critical('Conflicting input: --projection and --zoom can not be used together.')
        exit(1)
    
    if options.zoom is None and options.projection is None and options.scale is None:
        logging.critical('Bad geometry input: need at least one of --zoom, --scale, or --projection.')
        exit(1)
    
    geometry = get_geometry(options.projection, options.zoom, options.scale)
    
    #
    # Input and output files.
    #
    
    if not input_files:
        logging.critical('Missing input file(s).')
        optparser.print_usage()
        exit(1)
    
    if not (options.labels_file or options.places_file or options.registrations_file):
        logging.critical('Missing output file(s): labels, place points, or registration points.')
        optparser.print_usage()
        exit(1)
    
    #
    # Load places.
    #
    
    places = Places(bool(options.dump_file))
    
    for place in options.load_inputs(input_files, geometry, options.name_field, options.placement_field):
        places.add(place)
    
    #
    # Do the annealing.
    #
    
    annealer = Annealer(lambda p: p.energy, lambda p: p.move())

    if places.count() == 0:
        # looks like there's nothing to actually do.
        annealed = []
        
    elif options.temp_min and options.temp_max and options.steps:
        # soon-to-be-deprecated explicit temperatures-and-steps method
        annealed, e = annealer.anneal(places, options.temp_max, options.temp_min, options.steps, 30)
    
    elif options.processes == 1:
        # don't bother with multiprocessing when there's just the one
        annealed = anneal_in_serial(places)
    
    else:
        # use multiprocessing to work through pieces in parallel
        annealed = anneal_in_parallel(places, options.processes)
            
    #
    # Output results.
    #
    
    label_data = dict(type='FeatureCollection', features=[])
    place_data = dict(type='FeatureCollection', features=[])
    rgstr_data = dict(type='FeatureCollection', features=[])
    
    placed = FootprintIndex(geometry)
    
    for place in annealed:
        blocker = placed.blocks(place)
        overlaps = bool(blocker)
        
        if blocker:
            logging.info('%s blocked by %s' % (place.name, blocker.name))
        else:
            placed.add(place)
        
        properties = copy(place.properties)
        
        if options.include_overlaps:
            properties['overlaps'] = int(overlaps) # 1 or 0
        elif overlaps:
            continue
        
        #
        # Output slightly different geometries depending
        # on whether we want projected or geographic output.
        #
        
        label_feature = {'type': 'Feature', 'properties': properties}
        point_feature = {'type': 'Feature', 'properties': properties}

        label_feature['geometry'] = {'type': 'Polygon', 'coordinates': None}
        point_feature['geometry'] = {'type': 'Point', 'coordinates': None}

        reg_point, justification = place.registration()

        if options.output_projected:
            label_coords = list(place.label().envelope.exterior.coords)
    
            label_feature['geometry']['coordinates'] = label_coords
            label_data['features'].append(label_feature)
    
            point_feature['geometry']['coordinates'] = [place.position.x, place.position.y]
            place_data['features'].append(deepcopy(point_feature))
            
            point_feature['geometry']['coordinates'] = (reg_point.x, reg_point.y)
            
        else:
            lonlat = lambda xy: geometry.point_lonlat(xy[0], xy[1])
            label_coords = [map(lonlat, place.label().envelope.exterior.coords)]
    
            label_feature['geometry']['coordinates'] = label_coords
            label_data['features'].append(label_feature)
    
            point_feature['geometry']['coordinates'] = [place.location.lon, place.location.lat]
            place_data['features'].append(deepcopy(point_feature))
            
            point_feature['geometry']['coordinates'] = lonlat((reg_point.x, reg_point.y))

        point_feature['properties']['justified'] = justification
        rgstr_data['features'].append(point_feature)
    
    #
    # Write to output files
    #
    
    datas = label_data, place_data, place_data
    files = options.labels_file, options.places_file, options.registrations_file
    
    for (data, filename) in zip(datas, files):
        if not filename:
            continue
    
        if options.append and exists(filename):
            previous_data = json.load(open(filename))
            data['features'] = previous_data['features'] + data['features']
        
        json.dump(data, open(filename, 'w'), indent=2)
    
    if options.dump_file:
        frames = []
        
        while places.previous:
            current, places = places, places.previous
            current.previous = None # don't pickle too much per state
            frames.append(current)
        
        frames = [frames[i] for i in range(0, len(frames), options.dump_skip)]
        frames.reverse()
        
        logging.info('Pickling %d states to %s' % (len(frames), options.dump_file))
        
        cPickle.dump(frames, open(options.dump_file, 'w'))
